File: 2018/day5/part2.rb
input = File.read("input.txt").chomp.chars

lengths = {}
result = ('a'..'z').min_by do |letter|
  new_input = input.reject {|c| c.downcase == letter}

  result = new_input.reduce("") do |acc, x|
    last = acc[-1] || ""
    if last.downcase == x.downcase && (last.ord - x.ord).abs == 32
      acc[0...-1]
    else
      acc + x
    end
  end
  lengths[letter] = result.length
end
puts lengths[result]

File: 2018/day5/part2.exs
chars =
  File.read!("input.txt")
  |> String.trim()
  |> String.to_charlist()

?a..?z
|> Enum.map(fn letter ->
  chars
  |> Enum.reject(fn x -> letter == x || letter == x + 32 end)
  |> Enum.reduce([], fn
    x, [] -> [x]
    x, [h | t] when abs(x - h) == 32 -> t
    x, acc -> [x | acc]
  end)
  |> length
end)
|> Enum.min()
|> IO.inspect()
